File: apps/android/app/src/main/java/com/futo/notes/EditorDraftCoordinator.kt
package com.futo.notes

/**
 * Admission control for fire-and-forget editor draft flushes.
 *
 * Rename, move, and delete advance an identity generation synchronously before
 * suspending. Flushes admitted under an older generation are rejected, while
 * the identity mutation waits for the previously admitted queue tail.
 */
internal class EditorDraftCoordinator {
    data class Admission(val id: String, val generation: Long, val epoch: Long = 0)
    data class IdentityMutation(val id: String, val generation: Long, val epoch: Long)

    private var epoch = 0L
    private var paused = false

    @Synchronized
    fun beginReset() { paused = true; epoch++ }

    @Synchronized
    fun endReset() { paused = false; generations.clear(); blocked.clear() }

    private val generations = mutableMapOf<String, Long>()
    private val blocked = mutableSetOf<String>()

    @Synchronized
    fun admit(id: String): Admission? =
        if (paused || id in blocked) null else Admission(id, generations[id] ?: 0, epoch)

    @Synchronized
    fun permits(admission: Admission): Boolean =
        !paused && admission.epoch == epoch && admission.id !in blocked &&
            (generations[admission.id] ?: 0) == admission.generation

    @Synchronized
    fun beginIdentityMutation(id: String): IdentityMutation {
        val generation = (generations[id] ?: 0) + 1
        generations[id] = generation
        blocked += id
        return IdentityMutation(id, generation, epoch)
    }

    @Synchronized
    fun finishIdentityMutation(mutation: IdentityMutation, committed: Boolean) {
        if (mutation.epoch != epoch || generations[mutation.id] != mutation.generation) return
        if (!committed) blocked -= mutation.id
    }

    /** A newly created/written note at a previously deleted id is a new owner. */
    @Synchronized
    fun reopen(id: String) {
        if (id !in blocked) return
        generations[id] = (generations[id] ?: 0) + 1
        blocked -= id
    }
}

File: apps/ios/Sources/Editor/EditorDraftCoordinator.swift
import Foundation

/// Admission control for fire-and-forget editor draft flushes. Identity
/// mutations advance a generation before suspending, so queued old-identity
/// flushes cannot recreate a deleted, renamed, or moved note.
@MainActor
final class EditorDraftCoordinator {
    struct Admission: Equatable {
        let id: String
        let generation: UInt64
        let epoch: UInt64
    }

    struct IdentityMutation: Equatable {
        let id: String
        let generation: UInt64
        let epoch: UInt64
    }

    private var epoch: UInt64 = 0
    private var paused = false

    func beginReset() {
        paused = true
        epoch += 1
    }
    func endReset() {
        paused = false
        generations.removeAll()
        blocked.removeAll()
    }

    private var generations: [String: UInt64] = [:]
    private var blocked: Set<String> = []

    func admit(_ id: String) -> Admission? {
        guard !paused, !blocked.contains(id) else { return nil }
        return Admission(id: id, generation: generations[id, default: 0], epoch: epoch)
    }

    func permits(_ admission: Admission) -> Bool {
        !paused && admission.epoch == epoch && !blocked.contains(admission.id)
            && generations[admission.id, default: 0] == admission.generation
    }

    func beginIdentityMutation(_ id: String) -> IdentityMutation {
        let generation = generations[id, default: 0] + 1
        generations[id] = generation
        blocked.insert(id)
        return IdentityMutation(id: id, generation: generation, epoch: epoch)
    }

    func finishIdentityMutation(_ mutation: IdentityMutation, committed: Bool) {
        guard mutation.epoch == epoch, generations[mutation.id] == mutation.generation else {
            return
        }
        if !committed { blocked.remove(mutation.id) }
    }

    func reopen(_ id: String) {
        guard blocked.contains(id) else { return }
        generations[id, default: 0] += 1
        blocked.remove(id)
    }
}
